npm/pkg/dataplane/dataplane.go

```diff
@@ -284,15 +284,29 @@ func (dp *DataPlane) RemovePolicy(policyKey string) error {
 	// because policy Manager will remove from policy from cache
 	// keep a local copy to remove references for ipsets
 	policy, ok := dp.policyMgr.GetPolicy(policyKey)
+	endpoints := policy.PodEndpoints
 	if !ok {
 		klog.Infof("[DataPlane] Policy %s is not found. Might been deleted already", policyKey)
 		return nil
 	}
+
+	dp.updatePodCache.Lock()
+
 	// Use the endpoint list saved in cache for this network policy to remove
 	err := dp.policyMgr.RemovePolicy(policy.PolicyKey)
 	if err != nil {
 		return fmt.Errorf("[DataPlane] error while removing policy: %w", err)
 	}
+
+	for npmEndpoint := range endpoints {
+		// if the endpoint is not in the policy's endpoint list, delete from cache
+		if _, ok := policy.PodEndpoints[npmEndpoint]; !ok {
+			delete(dp.endpointCache.cache, npmEndpoint)
```

Comment:
general structure is good, but we're looking to delete the policy reference from the given endpoint the endpoint cache is also indexed by IP

The endpoint cache is also indexed by IP. If Policy.Endpoints doesn't have an IP, then the lookup may not be efficient but that's ok

Reply:
updated